Map Refactor

dwatchguard opened the pull request:
NEW CLASSES:
Grid, GridMap, PathMap, MapHelper, AttackHelper, logic_pair, logic_reference_wrapper, ID, 

REMOVED CLASSES:
Map, CellPath, Tile

Map has now been renamed GridMap and extends Grid. 
MoveHelper and AttackHelper now uses actual Ranges from a Unit's Items
MapHelper can find all Cells within any Range of other Cells

GridCells are now identified by an ID that only the Grid can use to determine what GridCell it represents.


Resolves: #19 Resolves: #16 Resolves #27

dwatchguard committed:
Tile no longer holds Unit*, Map renamed to Grid, Grid now holds map from GridCell to Unit

dwatchguard committed:
Created GridMap, a super class of Grid, and moved all Party and shortestPath code fiton it from Grid.

dwatchguard committed:
GridCell& can now be pulled from Grid using[index][index], removed getCell(x, y)

dwatchguard committed:
Updated GridMap to return constPathMap on const calls.

dwatchguard committed:
Created ID class to hold a GridCell's ID that can only be deciphered by the map. Now used in PathMap.

Added unused functionaliity  for a PathMap wrapper

Created MapHelper class that future helpers will extend (Will need to give these a more descriptive name)

Can now search the map using Ranges in MapHelper.

TerrainFactory Method now works as intended.

dwatchguard committed:
Added more operations to ID

dwatchguard committed:
Comment error

dwatchguard committed:
Merge Iterator change and more from master

dwatchguard committed:
Fix for GridMap::getParty(Unit)

dwatchguard committed:
PathMap now wraps an map<id, <int, id>>, that only gives access to CellRef, instad of the underlying ID, while maintaining const-ness

dwatchguard committed:
Implemented the logic_pair so that std:pairs can be better used in data structures. Placed path_map_iter into the Path_Map class

dwatchguard committed:
Removed Tile.cpp due to redundency.
Gave CridCell a terrain.
Clean-up magic numbers

dwatchguard committed:
Fixed bugs on when accessing PathMap elements

dwatchguard commented on Project1/PathMap.h:
```diff
@@ -3,15 +3,16 @@
+
 //TODO(Torrey): rename this class
-template <typename grid>
+template <typename grid, typename cell>
```
I feel like there is a better way than also having the cell in the template, but I don't know how

dwatchguard committed:
Typo Fix

dwatchguard committed:
Used Templates to replace const_grid_row in Grid

dwatchguard committed:
CellEdge now hold an ID instead of the corresponding GridCell&. This keeps Grid information more secure.

SilensAngelusNex commented on Project1/CellEdge.cpp:
```diff
@@ -21,7 +21,7 @@ std::optional<int> CellEdge::getCost(MobilitySet mobility_type, bool intangible)
 	for (MobilityType mobility : MobilityType::list) {
 		if (mobility_type[mobility]) {
 			std::optional<int> edge_cost = getCost(mobility);
-			if (edge_cost.has_value() && canPass(intangible)) { // if we can step on the tile
```
What happened to canPass?

dwatchguard replied:
CanPass was moved from edge to map, since edges no longer hold Units

SilensAngelusNex commented on Project1/CellPath.h:
```diff
@@ -1,19 +1,19 @@
+#include "GridCell.h"
+#include "ID.h"
 
+//TODO(Torrey): Create Superclasses of this. EX: UnitPath has a unit to traverse the Grid.
```
subclasses?

dwatchguard replied:
Potatoe Potatoe

SilensAngelusNex commented on Project1/Grid.h:
```diff
@@ -0,0 +1,76 @@
+public:
+	//vector wrapper that only allows [] op
+	template<typename VecCell>
+	class grid_row {
```
This, a view of a vector that you cannot add or remove elements from, should be its own standalone class and also should expose more methods of vector than just operator[].

dwatchguard replied:
check

SilensAngelusNex commented on Project1/GridMap.cpp:
```diff
@@ -0,0 +1,148 @@
+}
+
+PathMap GridMap::findShortestPaths(size_t x, size_t y) {
+	return PathMap(*this, getShortestPaths((*this)[x][y]._id, INT_MAX, MobilitySet(MobilityType::values::GROUNDED), false));
```
Since operator[] is public, I would remove this method and make the caller find the id they want.

And I haven't been doing this either, but [default arguments](https://www.geeksforgeeks.org/default-arguments-c/) are a thing if you want to try that out.

dwatchguard replied:
Agre on this, it seems like it is needlessly verbose and useless

dwatchguard replied:
check

SilensAngelusNex commented on Project1/Grid.h:
```diff
@@ -0,0 +1,76 @@
+	bool hasUnit(const Unit & unit) const;
+
+	Unit& operator[](const GridCell& index);
+	GridCell& operator[](const Unit& index);
```
I think this should return an ID

dwatchguard replied:
I will need to think this through, that might be a bit messy to change

dwatchguard replied:
I did this, but I think I want to rethink  what the [] operator means

SilensAngelusNex commented on Project1/Grid.h:
```diff
@@ -0,0 +1,76 @@
+	bool hasUnit(const GridCell & index) const;
+	bool hasUnit(const Unit & unit) const;
+
+	Unit& operator[](const GridCell& index);
```
I think this should be a named method that takes an ID

Probably all of the public methods that take `const GridCell&` you should consider eitther taking an ID instead or overloading for ID.

SilensAngelusNex commented on Project1/Grid.h:
```diff
@@ -0,0 +1,76 @@
+	void insertUnit(Unit& new_unit, GridCell& destination);
+	void removeUnit(Unit& unit);
+	
+	bool hasUnit(const GridCell & index) const;
```
Takes an ID?

SilensAngelusNex commented on Project1/GridMap.cpp:
```diff
@@ -0,0 +1,148 @@
+/////////////////////////////////////////////////////////////////////////
+
+
+id_cost_map GridMap::getShortestPaths(const Unit& unit) const {
```
Do we want to try to refactor Unit/Mobility so that this function can call the other?

SilensAngelusNex commented on Project1/ID.h:
```diff
@@ -0,0 +1,101 @@
+
+};
+
+class ID :
```
Is this not just:
```
class ID : private logic_pair<size_t, size_t> {
public:
    using logic_pair<size_t, size_t> operator==;
    using logic_pair<size_t, size_t> operator!=;
    using logic_pair<size_t, size_t> operator<=;
    using logic_pair<size_t, size_t> operator>=;
    using logic_pair<size_t, size_t> operator<;
    using logic_pair<size_t, size_t> operator>;
    friend class GridCell;
}
```

dwatchguard replied:
I am sure I can find a way to use auto or decltype to fix this, but no. logic_pair< takes a logic_pair right operand, which this can't convert to.

SilensAngelusNex replied:
[Might want to take a look at this.](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) I'm not sure if it solves the problem, but it looks like it'll be useful elsewhere if it isn't here.

I think you'd declare logic pair like:
```
template<typename A, typename B, typename Derived = logic_pair<A, B>>
class logic_pair {
    bool operator>=(const Derived& other) const {
        return whatever;
    }

```

dwatchguard replied:
I tried that and it did not compile. Im kinda fine with how ID is right now, since I might make Grid and GridMap and ID have a template argument in the future to better support different map shapes, so ID may not always be a logic_pair anyway

SilensAngelusNex commented on Project1/MapHelper.h:
```diff
@@ -0,0 +1,35 @@
+	std::vector<UnitRef> getAlliesWithin(const Unit& unit, Range range);
+	std::vector<UnitRef> getEnemiesWithin(const Unit& unit, Range range);
+
+	std::vector<UnitRef> getUnitsWithin(Range range, GridCell& pos);
```
IDs?

SilensAngelusNex commented on Project1/MoveHelper.h:
```diff
@@ -1,40 +1,42 @@
-	Map& _map;
 	template<typename T>
-	void vectorSubtract(std::vector<T>& a, std::vector<T>& b) {
+	void vectorSubtract(std::vector<std::reference_wrapper<T>>& a, std::vector<std::reference_wrapper<T>>& b) {
```
this should be a free function

dwatchguard replied:
Also, I couldn't find a one line set difference method in the STL to do this, so I made one...

dwatchguard replied:
it also is free...

SilensAngelusNex commented on Project1/PathMap.h:
```diff
@@ -0,0 +1,113 @@
+template <typename grid, typename cell>
+class Path_Map :
+	private id_cost_map
+{
```
`class Path_Map : private id_cost_map {`

Please, sir. Newlines don't grow on trees around here.

dwatchguard replied:
The Lorax is sad

SilensAngelusNex replied:
2 lines is one less than 3, but still twice as many as 1.

dwatchguard replied:
that is outdated westy

SilensAngelusNex commented on Project1/Range.h:
```diff
@@ -19,6 +20,19 @@ class Range {
 	Range(MobilitySet move, DistanceSet dist);
 
+	constexpr bool hasDistance(int dist) const { 
+		Expects(dist >= 0 && MAX_RANGE > dist);
```
`&& MAX_RANGE * MOVE_RATIO > dist`

dwatchguard replied:
check

SilensAngelusNex commented on Project1/Unit.h:
```diff
@@ -108,4 +107,8 @@ class Unit : public Observable<ObserverDamage>, public Observable<ObserverExp> {
-};
\ No newline at end of file
+};
+//TODO (Weston): Is this an ok place for this?
```
I think these should go inside of classes, ie
`Unit::Ref`
`Unit::ConstRef`
`Unit::UniquePtr`

dwatchguard replied:
check

SilensAngelusNex commented on Project1/reference_logic.h:
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
```
Why? This is scary, because reference_wrapper is implicitly convertible to reference, so it makes sense to do `a == b` to try to get value equals, but you're going to get pointer equals instead.

dwatchguard replied:
Just Commenting to remind myself that I fixed this

SilensAngelusNex commented on Project1/resource.h:
```diff
@@ -0,0 +1,14 @@
+// Used by Game.rc
+
+// Next default values for new objects
+// 
```
Uhhh what?

SilensAngelusNex reviewed (commented):
Here you go!

dwatchguard committed:
Temporarily completely removed CellPath to better build from the ground up

Added small fixes Weston mentioned

dwatchguard committed:
Used Class Namespaces to rename CellRef and UnitRef

dwatchguard committed:
Find ShortestPaths now uses lambda expressions to determine if it can cross a Cell.

More places have ID overrides instead of GridCell ones

GridCell boolean operators have received an upgrade.

dwatchguard committed:
Grid[Unit] now returns an ID
Grid now has more conversion getters to account for this.

dwatchguard committed:
Added read_vector.h to make non modifieable vectors, and used it to refactor _grid_row

dwatchguard committed:
Clang and `using` renames

dwatchguard committed:
Removed Extraneous getCost method, and fixed getCost min cost logic

dwatchguard committed:
BasicWeapon::canEquip(Unit&) is now BasicWeapon::canEquip(const Unit&)

Weapon Range Calculations now use attack range of equipped main hand weapon.

Added Helper function to InventoryViewable to check if there is a weapon equiped in a given slot

dwatchguard committed:
MoveHelper now uses ID over GridCell::Ref

dwatchguard committed:
AttackHelper, MapHelper, and MoveHelper backbone

dwatchguard committed:
AttackHelper Now can getAttackableUnits, and can return if two units are in position to engage in combat.

Refereve_Wrapper boolean logi was enhanced

dwatchguard committed:
Cretated logical_reference_wrapepr to handle all of our reference_wrapped comparison needs.

dwatchguard committed:
Added functionality to find where a Unit can assist
Used Templates for ATtack/Assist ranges to prevent code dupe

dwatchguard committed:
Added Templates to make code easier to maintain

dwatchguard committed:
Added const-correctness for Map and Move helper

dwatchguard committed:
Moved some Template methods to cpp files

dwatchguard committed:
Moved more template functions to cpp files

dwatchguard committed:
Clang Compiled/ Tidied
Fixed bug where certain MoveHelper Fuctions returned the wrong sets,
Improved Const corectness

dwatchguard committed:
Merge branch 'master' into map_refactor

dwatchguard committed:
Removed reference_logic.h

dwatchguard committed:
Cretated id_cost_map to hold an origin, and renamed Path_Map to path_map

SilensAngelusNex commented on Project1/AttackHelper.h:
```diff
@@ -0,0 +1,23 @@
+	AttackHelper(GridMap& map);
+
+	std::vector<Unit::Ref> getAttackableUnits(const Unit& unit);
+	const std::vector<Unit::Ref> getAttackableUnits(const Unit& unit) const;
```
I don't think the constness here makes sense. If you're trying to keep callers of the const method from modifying the units, I don't think it works. They can just create a new vector with all the unit refs in it and modify them from there.

If that's what you want, the const one should return a vector of const Unit::Red

dwatchguard replied:
I don't think it makes sense to have a non-const version anyway, since this is for combat and attacking is not const. I'll just remove it

SilensAngelusNex commented on Project1/AttackHelper.cpp:
```diff
@@ -0,0 +1,33 @@
+	if (unit.getInventory().hasEquip(EquipSlot::values::ON_HAND)) {
+		range = unit.getInventory()[EquipSlot(EquipSlot::values::ON_HAND)].getAttackRange();
+	}
+	return getEnemiesWithin(range, _map[unit], *unit.getParty());
```
```
if (unit.getInventory().hasEquip(EquipSlot::values::ON_HAND)) {
	range = unit.getInventory()[EquipSlot(EquipSlot::values::ON_HAND)].getAttackRange();
	return getEnemiesWithin(range, _map[unit], *unit.getParty());
}
return {};
```

No sense in making the function call if you know it's going to return empty.

SilensAngelusNex commented on Project1/Grid.cpp:
```diff
@@ -29,13 +29,29 @@ Grid::Grid(int width, int height, std::vector<PartyData> data) {
+Unit* Grid::getUnit(const ID & index) {
+	return  hasUnit(index) ? _cell_to_unit.at(&(*this)[index]) : nullptr;
+}
+GridCell * Grid::getCell(const Unit & index)
```
parameter named index?

dwatchguard replied:
I use it as an index so...

dwatchguard replied:
key might be a better name :/

SilensAngelusNex commented on Project1/Grid.h:
```diff
@@ -60,16 +44,32 @@ class Grid {
+	bool hasUnit(const ID& index) const;
+
+	Unit* getUnit(const GridCell& index);
+	Unit* getUnit(const ID& index);
```
should probably take IDs by value since they're just int pairs

dwatchguard replied:
they are _size_t_ pairs thank you very much

SilensAngelusNex commented on Project1/GridMap.cpp:
```diff
@@ -33,116 +61,46 @@ const Party& GridMap::getParty(const Unit & unit) const {
-PathMap GridMap::findShortestPaths(ID start, int max_move, MobilitySet mobility, bool intangible) {
-	return PathMap(*this, getShortestPaths(start, max_move, mobility, intangible));
+PathMap GridMap::getShortestPathsMap(ID start, int max_move, MobilitySet mobility, bool intangible) {
+	return PathMap( *this, getShortestPathsHelper(start, max_move, mobility, [intangible](const Unit* unit) { return intangible || unit == nullptr; }, *this));
```
You can avoid having to pass `*this` as an argument to getShortestPathsHelper by capturing it in this lambda:
`[intangible, *this](const CellEdge& edge) { return intangible || getUnit(edge.)id) == nullptr; }`
and in the other lambda:
`[&unit, *this](const CellEdge& edge) { return unit.getMobility().canPass(getUnit(edge.)id)); }`

then you can just call `canPass(edge)` in the helper

dwatchguard replied:
getShortestPathsMap(...) needs access to the grid though...

SilensAngelusNex commented on Project1/MapHelper.h:
```diff
@@ -4,32 +4,51 @@
+	std::vector<GridCell::Ref> getAlliedCellsWithin(Range range, ID pos, const PartyBase& party);
+	std::vector<GridCell::Ref> getEnemyCellsWithin(Range range, ID pos, const PartyBase& party);
+
+	const std::vector<Unit::Ref> getUnits() const;
```
same thing as above with the const vectors; I don't think this does what you want.

SilensAngelusNex commented on Project1/logic_pair.h:
```diff
@@ -0,0 +1,50 @@
+		return first != other.first || second != other.second;
+	}
+	template<typename pair>
+	bool operator<(const pair& other) const {
```
Nitpick: These don't short-circuit and it would be nice if they did

Also, you should be using logical-or `||`, not bitwise-or `|`

dwatchguard replied:
short circuit?

dwatchguard replied:
Doing that hurts my brain, but I think I 'fixed' it

SilensAngelusNex commented on Project1/id_cost_map.h:
```diff
@@ -0,0 +1,59 @@
+#include "ID.h"
+#include "logic_pair.h"
+#include "GridCell.h"
+class id_cost_map : private std::map<ID, CostID>
```
why private inheritance if you expose all the methods?

dwatchguard replied:
I was not going to do them all, but caved in is why

dwatchguard committed:
Fixed review issues other than ID
Improved formatting

dwatchguard committed:
Fixed typo and PathMap [] logic

dwatchguard committed:
ID logic changes

dwatchguard committed:
Syntax Error, renamed parameters in Grid getters

dwatchguard committed:
Finally fixed <= and >= in logic_pair
